pkg/config/nodetreemodel/inner_node.go

```diff
@@ -0,0 +1,190 @@
+}
+
+// GetChild returns the child node at the given case-insensitive key, or an error if not found
+func (n *innerNode) GetChild(key string) (Node, error) {
+	mkey := n.remapCase[strings.ToLower(key)]
+	child, found := n.val[mkey]
+	if !found {
+		return nil, ErrNotFound
+	}
+	return child, nil
+}
+
+// HasChild returns true if the node has a child for that given key
+func (n *innerNode) HasChild(key string) bool {
+	_, ok := n.val[key]
+	return ok
+}
+
+// Merge mergs src node within current tree
+func (n *innerNode) Merge(src InnerNode) error {
+	defer n.makeRemapCase()
+
+	for _, name := range src.ChildrenKeys() {
+		srcChild, _ := src.GetChild(name)
+
+		if !n.HasChild(name) {
+			n.val[name] = srcChild.Clone()
+		} else {
+			// We alredy have child with the same name
+
+			dstLeaf, dstIsLeaf := n.GetLeaf(name)
+			srcLeaf, srcIsLeaf := srcChild.(LeafNode)
+			if srcIsLeaf != dstIsLeaf {
+				return fmt.Errorf("tree conflict, can't merge inner and leaf nodes for '%s'", name)
+			}
+
+			if srcIsLeaf {
+				if srcLeaf.SourceGreaterOrEqual(dstLeaf.Source()) {
+					n.val[name] = srcLeaf.Clone()
+				}
+			} else {
+				dstInner, _ := n.GetInner(name)
+				childInner, _ := srcChild.(InnerNode)
+				if err := dstInner.Merge(childInner); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// ChildrenKeys returns the list of keys of the children of the given node, if it is a map
+func (n *innerNode) ChildrenKeys() []string {
+	mapkeys := maps.Keys(n.val)
+	// map keys are iterated non-deterministically, sort them
+	slices.Sort(mapkeys)
+	return mapkeys
+}
+
+// SetAt sets a value in the tree by either creating a leaf node or updating one if the priority is equal or higher than
+// the existing one. The function returns true if an update was done or false if nothing was changed.
+//
+// The key parts should already be lowercased.
+func (n *innerNode) SetAt(key []string, value interface{}, source model.Source) (bool, error) {
+	keyLen := len(key)
+	if keyLen == 0 {
+		return false, fmt.Errorf("empty key given to Set")
+	}
+
+	defer n.makeRemapCase()
+
+	part := key[0]
+	if keyLen == 1 {
+		node, ok := n.val[part]
+		if !ok {
+			n.val[part] = newLeafNodeImpl(value, source)
+			return true, nil
+		}
+
+		if leaf, ok := node.(LeafNode); ok {
+			if leaf.Source().IsGreaterOrEqualThan(source) {
+				n.val[part] = newLeafNodeImpl(value, source)
+				return true, nil
+			}
+			return false, nil
+		}
+		return false, fmt.Errorf("can't overrides inner node with a leaf node")
+	}
+
+	// new node case
+	if _, ok := n.val[part]; !ok {
+		newNode := newInnerNodeImpl()
+		n.val[part] = newNode
+		return newNode.SetAt(key[1:keyLen], value, source)
+	}
+
+	// update node case
+	node, ok := n.GetInner(part)
+	if !ok {
+		return false, fmt.Errorf("can't update a leaf node into a inner node")
+	}
+	return node.SetAt(key[1:keyLen], value, source)
+}
+
+// Set sets a node in the current node
+func (n *innerNode) Set(name string, node Node) {
```

Comment:
Would prefer a different name for this method, in order to avoid confusion with the `config.Set` user-facing API.